notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+      "2023-10-26 01:02:08.574353 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 2, 8, 574353)"
+      ]
+     },
+     "execution_count": 24,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "#MSQ async query using broadcast join\n",
+    "sql='''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, \n",
+    "  round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) as age_group, \n",
+    "  count(distinct c.client_ip) as distint_ips\n",
+    "FROM clicks c\n",
+    "     INNER JOIN users u\n",
+    "        ON c.user_id=u.user_id\n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) \n",
+    "ORDER BY 3 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "result = sql_client.async_sql(sql)\n",
+    "result=pandas.json_normalize(result.rows)\n",
+    "print_timestamp()\n",
+    "result"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "id": "2db4fe7e-47de-40eb-b3fa-2af7846a1638",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:11656.2219 ms   min:10813.459 ms  max:15201.207 ms\n",
+      "2023-10-26 01:04:05.147188 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 4, 5, 147188)"
+      ]
+     },
+     "execution_count": 25,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "print(measure_async_query(sql,10)) # using default broadcast join\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bdf9ad82-4098-4fab-bc7d-3b91068ee73e",
+   "metadata": {},
+   "source": [
+    "MSQ has an inherent disadvantage over Scatter/Gather when it comes to response time for many queries. It must first initialize the `query_controller` and `query_worker` tasks and only then begin to process the query.  You must also make sure that you have enough worker slots available for them. If any of the tasks are waiting to be assigned, the whole query will wait before it starts executing stages. It will fail after a timeout if worker slots do not become available.\n",
+    "MSQ reads the segment files directly from Deep Storage which is also typically slower than the local storage that Historicals use and there is no caching of segment files. It has the advantage that it can access data that is not cached in the historicals but still exists in Deep Storage. So it is best for longer running queries on broader or older timeframes.\n",
+    "\n",
+    "As queries are running you can see the `query_controller` and `query_worker` tasks being spawned for each query in the [tasks view of the Druid Console](http://localhost:8888/unified-console.html#tasks).\n"
```

Comment:
Is it possible to see this with an API call?

Reply:
The output is very hard to read. I think this is better.